AI correct/incorrect submission grouping

miguelaenlle opened the pull request:
Implementation of #12455, specialized for correct/incorrect submission grouping. User-ready version of #12524, which was used to prototype this feature.

Related PRs:
- #12684 
- #12685

---

When AI grading mode is enabled, users have the option to use AI submission grouping:

<img width="766" height="210" alt="Screenshot 2025-08-24 at 8 01 39 PM" src="https://github.com/user-attachments/assets/e470b8b6-a76a-433f-8fbd-82131a370c42" />

A correct answer must be specified in `pl-answer-panel` for submission grouping to work. For example,

<img width="334" height="140" alt="Screenshot 2025-08-24 at 8 08 08 PM" src="https://github.com/user-attachments/assets/1fd14d81-0de2-45d5-a65e-530b7449f3e4" />

Once the user clicks a one of the "Group submission" buttons, they'll see this modal:

<img width="846" height="719" alt="Screenshot 2025-08-24 at 8 02 44 PM" src="https://github.com/user-attachments/assets/9d645063-f2ab-4363-a2ec-0e2f15d9a75f" />

After a server job for submission grouping, the Submission Group column appears:

<img width="1080" height="422" alt="Screenshot 2025-08-24 at 8 03 14 PM" src="https://github.com/user-attachments/assets/ee828554-a535-47b5-a7b2-ba7f7df258ba" />

In the instance question grading view, the submission group appears as a field. The user can change the submission group -- modifications to the selected dropdown item automatically save.

<img width="374" height="494" alt="Screenshot 2025-08-24 at 8 06 56 PM" src="https://github.com/user-attachments/assets/70ff0472-da4e-452b-8482-44e8872d9d0a" />

Users also have the option to apply a grade to all submissions in the cluster (if the instance question is in a cluster).

<img width="437" height="283" alt="Screenshot 2025-08-24 at 8 07 36 PM" src="https://github.com/user-attachments/assets/b12c5748-6971-48a8-800b-35be15e1c0a6" />

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.html.tsx:
```diff
@@ -159,7 +162,7 @@ export function AssessmentQuestion({
                         <thead>
                           <tr class="table-light fw-bold">
-                            <td>Rubric item</td>
+                            <td style="width: 30%;">Rubric item</td>
```
Responsiveness? Prefer col?

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.sql:
```diff
@@ -77,6 +77,23 @@ SELECT
+  AND (
+    NOT $skip_graded_submissions
+    OR iq.requires_manual_grading
+  );
```
For now we only need to use the latest submission . That's ultimately what's graded anyways.

reteps commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ts:
```diff
@@ -136,6 +160,28 @@ router.post(
+
         res.redirect(res.locals.urlPrefix + '/jobSequence/' + jobSequenceId);
       } else {
         const action_data = JSON.parse(req.body.batch_action_data) || {};
```
I got a crash locally here, when I did "AI grade all submissions"

```
SyntaxError: \"undefined\" is not valid JSON\n    at JSON.parse (<anonymous>)\n    at <anonymous> (/Users/peter/work/PrairieLearn/apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ts:187:34)
```

miguelaenlle replied:
I wasn't able to reproduce the issue locally, but judging from the `SyntaxError`, this should resolve it:

https://github.com/PrairieLearn/PrairieLearn/blob/2d2493b5b52d63f6cacc36dd0ae88f4bd35f6cd2/apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ts#L193-L195

Let me know if this fixes it on your end.

reteps commented on apps/prairielearn/src/ee/lib/ai-submission-grouping/ai-submission-grouping-util.ts:
```diff
@@ -0,0 +1,113 @@
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const STANDARD_SUBMISSION_GROUPS = [
```
This isn't needed for my approval, but a UI issues for me is that:

<img width="1770" height="1292" alt="Image" src="https://github.com/user-attachments/assets/047112a1-f68b-41d8-bce0-6e109b70341f" />

- The submission count/average AI error is shown even when I haven't done AI grading
- The group explanations I probably only need to read once, so they take up a lot of space at the top of the page

I would even consider removing the group explanations in favor of an inline tooltip per column. The submission counts could maybe be hidden when AI grading hasn't been run yet.

reteps replied:
An actionable item: move the submission group descriptions to a tooltip in the column. On the individual question page, have a "?" tooltip that includes the submission group table on click/hover.

miguelaenlle replied:
Also, I agree with placing the submission group descriptions into tooltips in the column and in the instance question page, since it places that information closer to the actual submission group name and reduces clutter on the screen.

reteps replied:
We discussed putting a tooltip inside the cell explaining what the submission group meant, and removing the table from the top.

reteps commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/gradingPanel.html.ts:
```diff
@@ -92,6 +96,51 @@ export function GradingPanel({
               </li>
             `
           : ''}
+        ${showAiSubmissionGroup && context === 'main'
```
It confuses me that AI submission grouping is controlled by the AI grading mode toggle. I feel like the toggle might need a better name?

reteps replied:
I like both of those names.

reteps replied:
We discussed that this should happen in a follow-up PR.

reteps commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/gradingPanel.html.ts:
```diff
@@ -92,6 +96,51 @@ export function GradingPanel({
           : ''}
+        ${showAiSubmissionGroup && context === 'main'
+          ? html`
+              <li class="list-group-item align-items-center">
```
If I change the submission group, is there any way I see what the original group (from AI) was? I'm wondering if there is a way to indicate that the original value was overridden by a human.

<img width="560" height="416" alt="Image" src="https://github.com/user-attachments/assets/0cfbfe9c-78d7-4450-94e0-f3e7f4d2c30c" />

<img width="2490" height="858" alt="Image" src="https://github.com/user-attachments/assets/29e3755e-2af5-47d1-8f2d-67bf3415db97" />

In that same line of 'how do I restore the AI value?', is there any reason to have a 'group all ungrouped' option here?

<img width="526" height="386" alt="Image" src="https://github.com/user-attachments/assets/57c0915d-7ad7-439d-a3e2-02f94af6a33c" />

- I can't seem to filter by 'ungrouped' either

reteps commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/gradingPanel.html.ts:
```diff
@@ -152,32 +201,102 @@ ${submission.feedback?.manual}</textarea
+                id="next-instance-question-button"
                 class="btn btn-secondary"
-                href="${assessment_question_url}/next_ungraded?prior_instance_question_id=${resLocals
+                href="${assessment_question_url}/next?prior_instance_question_id=${resLocals
```
When I use this button with a single question not in a submission group, and hit next, I get kicked out to the main screen. Is the behavior such that it does "next submission in submission group"? Can that be turned off?

<img width="884" height="1236" alt="Image" src="https://github.com/user-attachments/assets/dc9bd622-ca14-4085-ac35-91bd2bff31ff" />

reteps replied:
TODO for me, no action needed: why is there no "previous" button?

reteps replied:
We discussed changing the text from 'Next' to 'Finish' on the last question in a group/overall.

miguelaenlle replied:
@nwalters512 , @mwest1066 and I discussed this over Zoom. Here are a few problems with showing a "Finish" button:
- A submission might come in while the instructor is grading, making "Finish" misleading. 
- Because the Next button does not necessarily loop back to all unseen submissions, implying that the user is about to finish is misleading. 

To address the first problem, we discussed flashing "All submissions graded" or showing a separate page with the same info, but neither addresses the second problem. We discussed that the best solution here is just leaving the functionality as-is, and addressing this as part of this issue: #12005

reteps commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/gradingPanel.html.ts:
```diff
@@ -152,32 +201,102 @@ ${submission.feedback?.manual}</textarea
+                              name="__action"
+                              value="add_manual_grade_for_submission_group_ungraded"
+                            >
+                              All ungraded instance questions in submission group
```
This feels like the wrong location for a bulk action. Should this live alongside the other bulk AI grading actions?

<img width="908" height="518" alt="Image" src="https://github.com/user-attachments/assets/cee96d2c-c58e-4c5e-9369-89a4911f8ab2" />

<img width="956" height="852" alt="Image" src="https://github.com/user-attachments/assets/a3b9f832-d6f8-4c76-a64e-b3052c2b54b5" />

miguelaenlle replied:
I agree with the intent of this. However, moving it to the dropdown in the assessment question page would necessitate creating a new screen with the grading panel and also not provide the instructor with a representative example of a submission in the cluster as they grade.

reteps replied:
Discussed that the previous implementation was signed off by Nathan and Matt, and doing this would be a major refactor, possible consideration for future work.

reteps commented on apps/prairielearn/src/ee/lib/ai-grading/ai-grading-util.ts:
```diff
@@ -181,19 +182,23 @@ export async function generatePrompt({
  * Parses the student's answer and the HTML of the student's submission to generate a message for the AI model.
  */
-function generateSubmissionMessage({
+export function generateSubmissionMessage({
```
nit: Can you add a JSDoc explaining why a caller would change `include_ai_grading_prompts` to `false`, and what the flag does?

miguelaenlle replied:
Added in https://github.com/PrairieLearn/PrairieLearn/commit/ecd96369c0773a69d4672ee2810c87b0c05234ac

reteps commented on apps/prairielearn/src/ee/lib/ai-grading/ai-grading-util.ts:
```diff
@@ -350,11 +357,14 @@ export function parseAiRubricItems({
 export async function selectInstanceQuestionsForAssessmentQuestion(
   assessment_question_id: string,
+  closed_instance_questions_only = false,
 ): Promise<InstanceQuestion[]> {
```
Nit: the return type for this is wrong. You should make your schema in this file, and use `z.infer<schema>` as the return type. See `db-types.ts` as an example.

miguelaenlle replied:
Added in ecd96369c0773a69d4672ee2810c87b0c05234ac

reteps commented on apps/prairielearn/src/ee/lib/ai-submission-grouping/ai-submission-grouping-util.ts:
```diff
@@ -0,0 +1,113 @@
+  );
+}
+
+export async function resetInstanceQuestionsAiSubmissionGroups({
```
nit: why is this called `reset` and not `delete`? `reset` implies it goes back to some known value IMO.

miguelaenlle replied:
Changed in ecd96369c0773a69d4672ee2810c87b0c05234ac

reteps commented on apps/prairielearn/src/lib/manualGrading.sql:
```diff
@@ -12,11 +33,26 @@ WITH
     WHERE
       iq.assessment_question_id = $assessment_question_id
       AND ai.assessment_id = $assessment_id -- since assessment_question_id is not authz'ed
+      AND (
```
Can you add an inline comment about this? I'm confused what this does.

reteps commented on apps/prairielearn/src/lib/manualGrading.ts:
```diff
@@ -41,23 +43,81 @@ const sql = sqldb.loadSqlEquiv(import.meta.url);
+  }
+
   // If we have no more submissions, then redirect back to main assessment question page
   return `${urlPrefix}/assessment/${assessment_id}/manual_grading/assessment_question/${assessment_question_id}`;
```
From a UI perspective: is there any reason we can't make the "Next" button e.g. read "finish" when you are about to be kicked out into the main page.

miguelaenlle replied:
This is doable; on the instance question page, I can call to retrieve the next instance question in order, and display "finish" if it doesn't exist.

reteps commented on apps/prairielearn/src/lib/manualGrading.sql:
```diff
@@ -1,4 +1,25 @@
--- BLOCK select_next_ungraded_instance_question
+-- BLOCK select_next_ai_submission_group_id
+SELECT
```
Just want to confirm how this interacts with the `next` system:

- 2 submissions in ai_submission_group_id 1
- 2 submissions in ai_submission_group_id 2

If I click next, starting with a submission in ai_submission_group_id 2, can I ever reach the submissions in ai_submission_group_id 1?

reteps replied:
In my opinion, we should order the way you navigate through the ai submission groups by using the highest instance_question_id in that group.

mwest1066 replied:
Also see https://github.com/PrairieLearn/PrairieLearn/issues/12005#issuecomment-3255363089 for recent discussion related to previous/next buttons.

reteps reviewed (commented):
Some initial high level thoughts. From a functionality perspective, everything seems to work pretty well. My main concern is about how submission groups interact with the "Next submission" feature.

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.html.tsx:
```diff
@@ -366,3 +446,121 @@ function DeleteAllAIGradingJobsModal({ csrfToken }: { csrfToken: string }) {
+  });
+}
+
+function GroupInfoModal({
```
That's correct. I'll include those as examples

miguelaenlle replied:
I added a handwritten submission example line in f8a105f45a690bf7c09d5e11f2b3e8843c0b451f

miguelaenlle commented on apps/prairielearn/src/migrations/20250915173901_instance_questions__ai_instance_question_group_id__manual_instance_question_group_id__add.sql:
```diff
@@ -0,0 +1,17 @@
+ALTER TABLE instance_questions
+ADD COLUMN manual_instance_question_group_id bigint REFERENCES instance_question_groups (id) ON UPDATE CASCADE ON DELETE SET NULL;
+
+ALTER TABLE instance_questions
+ADD CONSTRAINT instance_questions_ai_group_has_same_aq_id_fkey FOREIGN KEY (
+  assessment_question_id,
+  ai_instance_question_group_id
+) REFERENCES instance_question_groups (assessment_question_id, id) ON UPDATE CASCADE ON DELETE SET NULL;
+
+ALTER TABLE instance_questions
+ADD CONSTRAINT instance_questions_manual_group_has_same_aq_id_fkey FOREIGN KEY (
+  assessment_question_id,
+  manual_instance_question_group_id
+) REFERENCES instance_question_groups (assessment_question_id, id) ON UPDATE CASCADE ON DELETE SET NULL;
```
This is a good point. I'll remove the SET NULL in this place in particular, since the foreign key constraint on `ai_instance_question_group_id` will set it to NULL if the instance question group is removed.

miguelaenlle commented on apps/prairielearn/src/migrations/20250917044910_instance_questions__ai_instance_question_group_id__manual_instance_question_group_id__add.sql:
```diff
@@ -0,0 +1,23 @@
+ALTER TABLE instance_questions
+ADD CONSTRAINT instance_questions_ai_instance_question_group_id_fkey FOREIGN KEY (ai_instance_question_group_id) REFERENCES instance_question_groups (id) ON UPDATE CASCADE ON DELETE SET NULL NOT VALID;
+
+ALTER TABLE instance_questions
```
Add comments explaining these constraints and why they work

miguelaenlle replied:
Also test that they actually work

miguelaenlle replied:
Check documentation

reteps replied:
@miguelaenlle Are you planning to address these?

miguelaenlle replied:
Hi @reteps . Yes, I plan to address these. I clarified the comments above

miguelaenlle commented on apps/prairielearn/src/tests/accessibility/index.test.ts:
```diff
@@ -328,6 +328,7 @@ const SKIP_ROUTES = [
   // HTML documents.
   /^\/pl\/navbar\/course/,
   /^\/pl\/assessments_switcher\/course_instance/,
+  '/pl/course_instance/:course_instance_id/instructor/assessment/:assessment_id/manual_grading/instance_question/:instance_question_id/instance_question_groups/switcher',
```
@reteps , the line above it specifies that these are HTML fragments:

https://github.com/PrairieLearn/PrairieLearn/blob/ab84f28cdeee950bcbd5450c029c1e10cc56c8d2/apps/prairielearn/src/tests/accessibility/index.test.ts#L327-L328

Would you still recommend changing `HTML documents` to `HTML documents / fragments`?

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentQuestionClient.ts:
```diff
@@ -550,7 +586,12 @@ async function ajaxSubmit(this: HTMLFormElement, e: SubmitEvent) {
   const batchAction = formData.get('batch_action');
 
-  if (action === 'batch_action' && batchAction === 'ai_grade_assessment_selected') {
+  const newPageBatchActions = new Set([
```
I agree that this isn't an intuitive variable name. I changed it to `batchActions` in 6a16e0105ea9e39c0e530d43a7aa19ef04e1b424

miguelaenlle commented on apps/prairielearn/src/ee/lib/ai-grading/ai-grading-util.ts:
```diff
@@ -365,10 +377,12 @@ export function parseAiRubricItems({
 
 export async function selectInstanceQuestionsForAssessmentQuestion(
```
Made this change in 6a16e0105ea9e39c0e530d43a7aa19ef04e1b424

miguelaenlle commented on apps/prairielearn/src/ee/lib/ai-instance-question-grouping/ai-instance-question-grouping-util.ts:
```diff
@@ -0,0 +1,139 @@
+}: {
+  assessment_question_id: string;
+}) {
+  const hasGroups = await selectAssessmentQuestionHasInstanceQuestionGroups({
```
I updated this in `6a16e0105ea9e39c0e530d43a7aa19ef04e1b424`

https://github.com/PrairieLearn/PrairieLearn/blob/6a16e0105ea9e39c0e530d43a7aa19ef04e1b424/apps/prairielearn/src/ee/lib/ai-instance-question-grouping/ai-instance-question-grouping-util.ts#L27-L32

miguelaenlle commented on apps/prairielearn/src/lib/manualGrading.ts:
```diff
@@ -52,30 +55,84 @@ export async function nextInstanceQuestionUrl({
 }): Promise<string> {
-  const instance_question_id = await sqldb.queryOptionalRow(
+  const prior_instance_question_group_id = use_instance_question_groups
+    ? await run(async () => {
```
I updated this in 6a16e0105ea9e39c0e530d43a7aa19ef04e1b424

miguelaenlle replied:
I updated this:

https://github.com/PrairieLearn/PrairieLearn/blob/bed5a19fc43a29b2ba0d7fcd90ddfe17ccc9ee6d/apps/prairielearn/src/lib/manualGrading.ts#L68-L86

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ts:
```diff
@@ -97,6 +109,17 @@ router.get(
+
+    const use_instance_question_groups =
+      aiGradingMode &&
+      (await selectAssessmentQuestionHasInstanceQuestionGroups({
```
I made this change with a `run` statement to prevent unnecessary SQL queries if `aiGradingMode` is false:

https://github.com/PrairieLearn/PrairieLearn/blob/6a16e0105ea9e39c0e530d43a7aa19ef04e1b424/apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ts#L113-L123

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.sql:
```diff
@@ -77,6 +77,26 @@ SELECT
+    iq.ai_instance_question_group_id
+  ) = $selected_instance_question_group_id
+  AND ai.assessment_id = $assessment_id
+  AND (
+    NOT $skip_graded_submissions
+    OR iq.requires_manual_grading
+  );
```
Made this change in 6a16e0105ea9e39c0e530d43a7aa19ef04e1b424:

https://github.com/PrairieLearn/PrairieLearn/blob/6a16e0105ea9e39c0e530d43a7aa19ef04e1b424/apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.sql#L95-L99

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts:
```diff
@@ -396,12 +480,121 @@ router.post(
+      const aiGradingEnabled = await features.enabledFromLocals('ai-grading', res.locals);
+      const use_instance_question_groups =
+        aiGradingEnabled &&
+        (await selectAssessmentQuestionHasInstanceQuestionGroups({
```
I updated this in 6a16e0105ea9e39c0e530d43a7aa19ef04e1b424, as described in https://github.com/PrairieLearn/PrairieLearn/pull/12630#discussion_r2366665555

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts:
```diff
@@ -396,12 +480,121 @@ router.post(
+        : undefined;
+
+      for (const instanceQuestion of instanceQuestionsInGroup) {
+        const { modified_at_conflict } = await manualGrading.updateInstanceQuestionScore(
```
I agree with making that change, but in a separate PR since it affects many unrelated/loosely related features to AI submission grouping

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts:
```diff
@@ -396,12 +480,121 @@ router.post(
+          {
+            manual_score_perc: body.use_score_perc ? body.score_manual_percent : null,
+            manual_points: body.use_score_perc ? null : body.score_manual_points,
+            auto_score_perc: body.use_score_perc ? body.score_auto_percent : null, // maybe different
```
That was an unnecessary comment -- I removed it in 6a16e0105ea9e39c0e530d43a7aa19ef04e1b424

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/gradingPanel.html.ts:
```diff
@@ -93,6 +106,69 @@ export function GradingPanel({
+                      .id}/instructor/assessment/${resLocals.assessment
+                      .id}/manual_grading/instance_question/${resLocals.instance_question
+                      .id}/instance_question_groups/switcher"
+                    hx-trigger="mouseover once, focus once, show.bs.dropdown once delay:200ms"
```
@reteps In 936f0f01acfa964818205e68a3b77dc9140dd99e I adjusted `gradingPanel` to directly populate the dropdown options instead of using HTMX.

miguelaenlle commented:
@reteps I added the suggestions/fixes:

- Submission group filtering works now:

<img width="1019" height="444" alt="Screenshot 2025-09-25 at 11 29 08 PM" src="https://github.com/user-attachments/assets/9dbc85c1-73d3-4b00-a70a-a331f4ca907c" />


- Fixed the "grouping 1 instance question**s**" issue:

<img width="467" height="179" alt="Screenshot 2025-09-25 at 11 29 50 PM" src="https://github.com/user-attachments/assets/9efc101d-f75f-44c5-ac81-cd9848bcd92f" />


- For the AI grading mode toggle, over Zoom @nwalters512 and I discussed saving adjustments to it for a separate PR, since it also affects AI grading.

reteps commented:
Approval stands, 🚢 it!

nwalters512 reviewed (commented):
Database things look good, pending improved comments.

I have not looked at the majority of this PR, so I won't explicitly sign off on this. This is a big PR touching a lot of files, I'll trust that y'all have been diligent in reviewing and testing these changes.

miguelaenlle commented on apps/prairielearn/src/migrations/20250921045656_instance_question_groups__create.sql:
```diff
@@ -0,0 +1,17 @@
+-- This is needed to enforce that instance questions in the same
+-- instance question group have the same assessment_question_id
+ALTER TABLE instance_question_groups
+ADD CONSTRAINT instance_question_groups_unique_id_constraint UNIQUE (assessment_question_id, id);
```
I made a few extra updates to the comment in 65135f6c1fc18406e0f78c318f79f37e703d288a
